### PR TITLE
add python3 subpackage

### PR DESCRIPTION
**Pull request by pronix**
added changed for python3 subpackage and sime changes for code
based on http://fedora.portingdb.xyz/ - looks like require for pulp migration to python3

### CODE REVIEW

**Comment by frenzymadness** [comment]
Hello @pronix  

I'm working now on py3 compatibility for nectar but I have much more changes in the source code. Have you tried to run tests in both versions of Python? What result do you have?

I am now in situation when all tests passes in both version of Python but I cannot find some real life examples to test nectar functionality. Could somebody please give me some code snippets to test this module in both Pythons?

I also have some suggestions for changes in specfile, I will fill it in comments for specific lines.

Thank you for your help.

File: python-nectar.spec [hunk 1]
```diff
@@ -1,9 +1,11 @@
 %{!?python_sitelib: %global python_sitelib %(%{__python} -c "from %distutils.sysconfig import get_python_lib; print get_python_lib()")}
+%{!?python3_sitelib: %global python3_sitelib %(%{__python3} -c "from %distutils.sysconfig import get_python_lib; print(get_python_lib())")}
+%global sum A download library that separates workflow from implementation details
```
Comment (frenzymadness) [inline python-nectar.spec]: Maybe you can drop this global definition, put text to Summary: and next use %{summary} macro.

**Comment by pronix** [comment]
hi @frenzymadness 
tests for both version passed. i used flake8 and run-tets for that.

File: python-nectar.spec [hunk 2]
```diff
@@ -14,9 +16,8 @@ BuildRoot:      %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 BuildArch:      noarch
 
 BuildRequires:  python-setuptools
```
Comment (frenzymadness) [inline python-nectar.spec]: More specific python2-setuptools would be better IMHO

File: python-nectar.spec (hunk 1 above)
Comment (pronix) [inline python-nectar.spec]: i follow this instruction https://fedoraproject.org/wiki/Packaging:Python
show me other and i will change :)

File: python-nectar.spec [hunk 3]
```diff
@@ -26,24 +27,43 @@ such as the default "threaded" downloader, which uses the "requests" library
 with multiple threads. Other experimental downloaders have used tools like
 pycurl and eventlets.
 
+%package -n python2-nectar
```
Comment (frenzymadness) [inline python-nectar.spec]: hardcoded name here is not good, consider to add something like:

%global modname nectar

and

%package -n python2-%{modname}

File: python-nectar.spec [hunk 4]
```diff
@@ -26,24 +27,43 @@ such as the default "threaded" downloader, which uses the "requests" library
...
+Summary:        %{sum}
+Requires:       python-isodate >= 0.4.9
+Requires:       python-requests >= 2.4.3
+%description -n python2-nectar
+
```
Comment (frenzymadness) [inline python-nectar.spec]: Same for description which can have global definition and %{desc} macro in main package and all subpackages

File: python-nectar.spec [hunk 5]
```diff
@@ -26,24 +27,43 @@ such as the default "threaded" downloader, which uses the "requests" library
...
 
 %build
-%{__python} setup.py build
+%py2_build
+%py3_build
 
 %install
 rm -rf $RPM_BUILD_ROOT
 %{__python} setup.py install -O1 --skip-build --root %{buildroot}
+%{__python3} setup.py install -O1 --skip-build --root %{buildroot}
```
Comment (frenzymadness) [inline python-nectar.spec]: You can use smart macros for installation:

%py2_install
%py3_install

File: python-nectar.spec [hunk 6]
```diff
@@ -26,24 +27,43 @@ such as the default "threaded" downloader, which uses the "requests" library
...
-%files
+%files -n python2-nectar
 %{python_sitelib}/nectar/
 %{python_sitelib}/nectar*.egg-info
 %doc COPYRIGHT LICENSE.txt README.rst
 
+%files -n python3-nectar
+%{python3_sitelib}/nectar/
+%{python3_sitelib}/nectar*.egg-info
+%doc COPYRIGHT LICENSE.txt README.rst
```
Comment (frenzymadness) [inline python-nectar.spec]: License files should have own definition with specific macro:

%license LICENSE

File: python-nectar.spec (hunk 5 above)
Comment (pronix) [inline python-nectar.spec]: this macroses not include -O1 flag

**Comment by frenzymadness** [comment]
I downloaded you version and script run-tests.sh failed with following output. I used Python 3.5.1:

> $ ./run-tests.sh 
> E......Exception ignored in: <bound method DownloaderConfig.__del__ of <nectar.config.DownloaderConfig object at 0x7fab37246320>>
> Traceback (most recent call last):
>   File "/home/lbalhar/Software/nectar-pronix/nectar/config.py", line 186, in __del__
>     self.finalize()
>   File "/home/lbalhar/Software/nectar-pronix/nectar/config.py", line 196, in finalize
>     for file_name in self._temp_files:
> AttributeError: 'DownloaderConfig' object has no attribute '_temp_files'
> ..E..E.E
> ======================================================================
> ERROR: Failure: ImportError (No module named 'cStringIO')
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
>     raise self.exc_val.with_traceback(self.tb)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/loader.py", line 418, in loadTestsFromName
>     addr.filename, addr.module)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 47, in importFromPath
>     return self.importFromDir(dir_path, fqname)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 94, in importFromDir
>     mod = load_module(part_fqname, fh, filename, desc)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 234, in load_module
>     return load_source(name, filename, file)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 172, in load_source
>     module = _load(spec)
>   File "<frozen importlib._bootstrap>", line 693, in _load
>   File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 662, in exec_module
>   File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
>   File "/home/lbalhar/Software/nectar-pronix/test/unit/test_base_downloader.py", line 1, in <module>
>     from cStringIO import StringIO
> ImportError: No module named 'cStringIO'
> 
> ======================================================================
> ERROR: test_ssl_data_config_value (test_config.InstantiationTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/lbalhar/Software/nectar-pronix/test/unit/test_config.py", line 76, in test_ssl_data_config_value
>     config = DownloaderConfig(ssl_ca_cert=ca_cert_value)
>   File "/home/lbalhar/Software/nectar-pronix/nectar/config.py", line 136, in __init__
>     self._process_ssl_settings()
>   File "/home/lbalhar/Software/nectar-pronix/nectar/config.py", line 175, in _process_ssl_settings
>     os.write(data_arg_os_handle, data_arg_value)
> TypeError: a bytes-like object is required, not 'str'
> 
> ======================================================================
> ERROR: Failure: ImportError (No module named 'StringIO')
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
>     raise self.exc_val.with_traceback(self.tb)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/loader.py", line 418, in loadTestsFromName
>     addr.filename, addr.module)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 47, in importFromPath
>     return self.importFromDir(dir_path, fqname)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 94, in importFromDir
>     mod = load_module(part_fqname, fh, filename, desc)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 234, in load_module
>     return load_source(name, filename, file)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 172, in load_source
>     module = _load(spec)
>   File "<frozen importlib._bootstrap>", line 693, in _load
>   File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 662, in exec_module
>   File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
>   File "/home/lbalhar/Software/nectar-pronix/test/unit/test_local_downloader.py", line 7, in <module>
>     from StringIO import StringIO
> ImportError: No module named 'StringIO'
> 
> ======================================================================
> ERROR: Failure: ImportError (No module named 'cStringIO')
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
>     raise self.exc_val.with_traceback(self.tb)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/loader.py", line 418, in loadTestsFromName
>     addr.filename, addr.module)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 47, in importFromPath
>     return self.importFromDir(dir_path, fqname)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib/python3.5/site-packages/nose/importer.py", line 94, in importFromDir
>     mod = load_module(part_fqname, fh, filename, desc)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 234, in load_module
>     return load_source(name, filename, file)
>   File "/home/lbalhar/.virtualenvs/nectar3/lib64/python3.5/imp.py", line 172, in load_source
>     module = _load(spec)
>   File "<frozen importlib._bootstrap>", line 693, in _load
>   File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 662, in exec_module
>   File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
>   File "/home/lbalhar/Software/nectar-pronix/test/unit/test_threaded_downloader.py", line 1, in <module>
>     from cStringIO import StringIO
> ImportError: No module named 'cStringIO'
> 
> Name                Stmts   Miss  Cover
> nectar.py               0      0   100%
> nectar/config.py       75      6    92%
> nectar/request.py      18      8    56%
> TOTAL                  93     14    85%
> Ran 15 tests in 0.072s
> 
> FAILED (errors=4)

**Comment by pronix** [comment]
i will recheck evening. and if will required - i can provide server for tests or for ci. good idea - add project to some public ci

**Comment by frenzymadness** [comment]
Ok, I can make my own PR after some code review after porting. I have a lot of changes in source code but we will see if your code will work in Py3.

**Comment by pronix** [comment]
if it ready - let's make it. for me interesting porting pulp to python3. lazy support 2 version of python :)

**Comment by frenzymadness** [comment]
Hello.

With your latest version tests still don't work. I've just created my own PR with Py3 compatibility changes [0] (created mainly with futurize) but I cannot be completely sure that nectar is working with my changes because I haven't any real life example of use.

Could you please test my version or send me some examples to test it? Thank you very much.

[0] https://github.com/pulp/nectar/pull/51

**Comment by BrnoPCmaniak** [comment]
Any update on this PR?

**Comment by pronix** [comment]
as soon as i will find free time

**Comment by jeremycline** [comment]
Hi @pronix and @frenzymadness , you should know that in Pulp 3 (which is Python 3 only :sparkles:) we plan to drop Nectar and that we will retire this project, so it might be better to port other packages to Python 3. That being said, I really appreciate you working on this and contributing your time to making the Python world a better place! If you really still want to continue that's fine and I (or someone else with commit access) will review the PRs, but I think this is a very low-priority package to be converted.

**Comment by frenzymadness** [comment]
Hello @jeremycline.

Thank you for your message. Is there any planned release date for Pulp 3? Will it use the same name or will you rename it to *pulp3*?

**Comment by jeremycline** [comment]
There is not a planned release date. It will still be called pulp, so when it comes out downstream packages will be updated (so, for example, at some point the pulp-2.y package in Fedora Rawhide will be updated to pulp-3.0) rather than both being available at once.

**Pull request closed without merging**